core/commonMain/src/kotlinx/serialization/internal/ElementMarker.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2017-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.CompositeDecoder
+
+@OptIn(ExperimentalSerializationApi::class)
+@PublishedApi
+internal class ElementMarker(
+    private val descriptor: SerialDescriptor,
+    private val filter: (SerialDescriptor, Int) -> Boolean
+) {
+    /*
+    Element decoding marks from given bytes.
+    The element number is the same as the bit position.
+    Marks for the lowest 64 elements are always stored in a single Long value, higher elements stores in long array.
+     */
+    private var lowerMarks: Long
+    private val highMarksArray: LongArray
+
+    companion object {
+        val EMPTY_HIGH_MARKS = LongArray(0)
+    }
+
+    init {
+        val elementsCount = descriptor.elementsCount
+        if (elementsCount <= Long.SIZE_BITS) {
+            lowerMarks = if (elementsCount == Long.SIZE_BITS) {
+                // number of bits in the mark is equal to the number of fields
+                0L
+            } else {
+                // (1 - elementsCount) bits are always 1 since there are no fields for them
+                -1L shl elementsCount
+            }
+            highMarksArray = EMPTY_HIGH_MARKS
+        } else {
+            lowerMarks = 0L
+            highMarksArray = prepareHighMarksArray(elementsCount)
+        }
+    }
+
+    fun mark(index: Int) {
+        if (index < Long.SIZE_BITS) {
+            lowerMarks = lowerMarks or (1L shl index)
+        } else {
```

Comment:
Same comment here about splitting the code onto fast/slot paths, please consider extracting `else` branch content to a separate `markHigh`